polar-core/src/kb.rs

```diff
@@ -861,18 +854,30 @@ impl KnowledgeBase {
             }
         }
 
-        let mut rule_types = rule_types_to_create.into_iter().map(|((subject, relation_name, object), required)| {
+        let mut rule_types = rule_types_to_create.into_iter().map(|((subject, relation, object), required)| {
             let subject_specializer = pattern!(instance!(&subject.value().as_symbol().expect("must be symbol").0));
-            let relation_name = relation_name.value().as_string().expect("must be string");
+            let relation_name = relation.value().as_string().expect("must be string");
             let object_specializer = pattern!(instance!(&object.value().as_symbol().expect("must be symbol").0));
-            rule!("has_relation", ["subject"; subject_specializer, relation_name, "object"; object_specializer], required)
+
+            let src_id = relation.get_source_id().expect("must be parsed");
+            let (left, right) = relation.span().expect("must be parsed");
+
+            let mut params = args!("subject"; subject_specializer, relation_name, "object"; object_specializer);
+            params.reverse();
+            let body = term!(op!(And));
+            let mut rule = Rule::new_from_parser(src_id, left, right, sym!("has_relation"), params, body);
```

Comment:
We could add a `clone_with_value` method like we have for `Term` that carries across source info?